core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -118,7 +141,13 @@ class HadoopMapReduceCommitProtocol(
     }
 
     dir.map { d =>
-      new Path(new Path(stagingDir, d), filename).toString
+      if (dynamicPartitionOverwrite) {
+        val tempFile = new Path(dynamicStagingTaskPath(dir, taskContext), filename)
+        dynamicStagingTaskFiles += tempFile
+        tempFile.toString
+      } else {
+        new Path(new Path(stagingDir, d), filename).toString
+      }
```

Comment:
If the issue is caused by conflict file name, why is it only specific to `dynamicPartitionOverwrite`?